File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SpindexTest.java
package org.firstinspires.ftc.teamcode;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorEx;

import org.firstinspires.ftc.teamcode.Helper.Hardware;

@TeleOp(name="Spindex Test")
public class SpindexTest extends LinearOpMode {

    int TARGET_POCKET = 0;
    double MOTOR_POWER = 0.35;
    double GEAR_RATIO = 1.0;
    int OFFSET_TICKS = 0;
    int TOLERANCE_TICKS = 15;

    @Override
    public void runOpMode() {

        DcMotor spindex = Hardware.spindexMotor;

        spindex.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);

        spindex.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        spindex.setMode(DcMotor.RunMode.RUN_USING_ENCODER);

        double motorTicksPerRev = spindex.getMotorType().getTicksPerRev();
        double ticksPerSpindexRev = motorTicksPerRev * GEAR_RATIO;
        double ticksPerPocket = ticksPerSpindexRev / 3.0;

        telemetry.addData("motorTicksPerRev", motorTicksPerRev);
        telemetry.addData("ticksPerPocket", ticksPerPocket);
        telemetry.addData("GEAR_RATIO", GEAR_RATIO);
        telemetry.addData("OFFSET_TICKS", OFFSET_TICKS);
        telemetry.update();

        waitForStart();
        if (isStopRequested()) return;

        spindex.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        spindex.setMode(DcMotor.RunMode.RUN_USING_ENCODER);

        int pocket = ((TARGET_POCKET % 3) + 3) % 3;
        int targetTicks = (int) Math.round(OFFSET_TICKS + pocket * ticksPerPocket);

        spindex.setTargetPosition(targetTicks);
        spindex.setMode(DcMotor.RunMode.RUN_TO_POSITION);
        spindex.setPower(Math.abs(MOTOR_POWER));

        while (opModeIsActive()) {

            int current = spindex.getCurrentPosition();
            int target = spindex.getTargetPosition();
            int error = Math.abs(target - current);

            boolean atTarget = (error <= TOLERANCE_TICKS) && !spindex.isBusy();

            telemetry.addData("TARGET_POCKET", pocket);
            telemetry.addData("CurrentTicks", current);
            telemetry.addData("TargetTicks", target);
            telemetry.addData("ErrorTicks", error);
            telemetry.addData("AtTarget", atTarget);
            telemetry.update();

            if (atTarget) {
                spindex.setPower(0);
            }
        }

        spindex.setPower(0);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SpindexTest.kt
package org.firstinspires.ftc.teamcode

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorEx
import kotlin.math.abs

@TeleOp(name = "SpindexTest")
class SpindexTest : LinearOpMode() {
    var TARGET_POCKET: Int = 0
    var MOTOR_POWER: Double = 0.35
    var GEAR_RATIO: Double = 1.0
    var OFFSET_TICKS: Int = 0
    var TOLERANCE_TICKS: Int = 15

    override fun runOpMode() {
        val spindex = hardwareMap.get<DcMotorEx>(DcMotorEx::class.java, "spindex")

        spindex.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE)

        spindex.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER)
        spindex.setMode(DcMotor.RunMode.RUN_USING_ENCODER)

        val motorTicksPerRev = spindex.getMotorType().getTicksPerRev()
        val ticksPerSpindexRev = motorTicksPerRev * GEAR_RATIO
        val ticksPerPocket = ticksPerSpindexRev / 3.0

        telemetry.addData("motorTicksPerRev", motorTicksPerRev)
        telemetry.addData("ticksPerPocket", ticksPerPocket)
        telemetry.addData("GEAR_RATIO", GEAR_RATIO)
        telemetry.addData("OFFSET_TICKS", OFFSET_TICKS)
        telemetry.update()

        waitForStart()
        if (isStopRequested()) return

        spindex.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER)
        spindex.setMode(DcMotor.RunMode.RUN_USING_ENCODER)

        val pocket = ((TARGET_POCKET % 3) + 3) % 3
        val targetTicks = Math.round(OFFSET_TICKS + pocket * ticksPerPocket).toInt()

        spindex.setTargetPosition(targetTicks)
        spindex.setMode(DcMotor.RunMode.RUN_TO_POSITION)
        spindex.setPower(abs(MOTOR_POWER))

        while (opModeIsActive()) {
            val current = spindex.getCurrentPosition()
            val target = spindex.getTargetPosition()
            val error = abs(target - current)

            val atTarget = (error <= TOLERANCE_TICKS) && !spindex.isBusy()

            telemetry.addData("TARGET_POCKET", pocket)
            telemetry.addData("CurrentTicks", current)
            telemetry.addData("TargetTicks", target)
            telemetry.addData("ErrorTicks", error)
            telemetry.addData("AtTarget", atTarget)
            telemetry.update()

            if (atTarget) {
                spindex.setPower(0.0)
            }
        }

        spindex.setPower(0.0)
    }
}
